File: samples/java/SupportChat/Server/Configuration/src/main/java/systems/zlink/samples/supportchat/server/configuration/SampleTopology.java
package systems.zlink.samples.supportchat.server.configuration;

import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("sample")
public record SampleTopology(
        String redisEndpoint,
        String redisKeyPrefix,
        String logDirectory,
        String apiChannelEndpoint,
        String apiSpotRouterEndpoint,
        String apiHttpEndpoint,
        String supportChannelEndpoint,
        String sessionStreamEndpoint,
        String sessionSpotRouterEndpoint,
        String supportSpotRouterEndpoint,
        String supportHttpEndpoint) {

    public Location location() {
        return new Location(
                required(redisEndpoint, "redisEndpoint"),
                required(redisKeyPrefix, "redisKeyPrefix"));
    }

    public String requiredLogDirectory() {
        return required(logDirectory, "logDirectory");
    }

    public Api api() {
        return new Api(
                required(apiChannelEndpoint, "apiChannelEndpoint"),
                required(apiSpotRouterEndpoint, "apiSpotRouterEndpoint"),
                required(apiHttpEndpoint, "apiHttpEndpoint"));
    }

    public Session session() {
        return new Session(
                required(sessionStreamEndpoint, "sessionStreamEndpoint"),
                required(sessionSpotRouterEndpoint, "sessionSpotRouterEndpoint"));
    }

    public Support support() {
        return new Support(
                required(supportChannelEndpoint, "supportChannelEndpoint"),
                required(supportSpotRouterEndpoint, "supportSpotRouterEndpoint"),
                required(supportHttpEndpoint, "supportHttpEndpoint"));
    }

    public static String configPath(String[] args) {
        if (args.length != 2 || !"--config".equals(args[0]) || args[1].isBlank()) {
            throw new IllegalArgumentException("Usage: <role executable> --config <path>");
        }
        return args[1];
    }

    private static String required(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("sample." + name + " is required");
        }
        return value;
    }

    public record Location(String redisEndpoint, String redisKeyPrefix) {}

    public record Api(String channelEndpoint, String spotRouterEndpoint, String httpEndpoint) {}

    public record Session(String streamEndpoint, String routerEndpoint) {}

    public record Support(String channelEndpoint, String routerEndpoint, String httpEndpoint) {}
}

File: samples/kotlin/SupportChat/Server/Configuration/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/configuration/SampleTopology.kt
package systems.zlink.samples.kotlin.supportchat.server.configuration

import org.springframework.boot.context.properties.ConfigurationProperties

@ConfigurationProperties("sample")
data class SampleTopology(
    val redisEndpoint: String?,
    val redisKeyPrefix: String?,
    val logDirectory: String?,
    val apiChannelEndpoint: String?,
    val apiSpotRouterEndpoint: String?,
    val apiHttpEndpoint: String?,
    val supportChannelEndpoint: String?,
    val sessionRouterEndpoint: String?,
    val supportSpotRouterEndpoint: String?,
    val streamEndpoint: String?,
) {
    fun location(): SampleLocation =
        SampleLocation(
            redisEndpoint = required(redisEndpoint, "redisEndpoint"),
            redisKeyPrefix = required(redisKeyPrefix, "redisKeyPrefix"),
        )

    fun logDirectory(): String = required(logDirectory, "logDirectory")

    fun api(): ApiTopology =
        ApiTopology(
            channelEndpoint = required(apiChannelEndpoint, "apiChannelEndpoint"),
            routerEndpoint = required(apiSpotRouterEndpoint, "apiSpotRouterEndpoint"),
            httpEndpoint = required(apiHttpEndpoint, "apiHttpEndpoint"),
        )

    fun session(): SampleSessionNode =
        SampleSessionNode(
            routerEndpoint = required(sessionRouterEndpoint, "sessionRouterEndpoint"),
            streamEndpoint = required(streamEndpoint, "streamEndpoint"),
        )

    fun support(): SupportTopology =
        SupportTopology(
            channelEndpoint = required(supportChannelEndpoint, "supportChannelEndpoint"),
            routerEndpoint = required(supportSpotRouterEndpoint, "supportSpotRouterEndpoint"),
        )

    companion object {
        fun configPath(args: Array<String>): String {
            require(args.size == 2 && args[0] == "--config" && args[1].isNotBlank()) {
                "Usage: <role executable> --config <path>"
            }
            return args[1]
        }

        private fun required(value: String?, name: String): String =
            requireNotNull(value?.takeIf { it.isNotBlank() }) { "sample.$name is required" }
    }
}

data class SampleLocation(val redisEndpoint: String, val redisKeyPrefix: String)

data class ApiTopology(
    val channelEndpoint: String,
    val routerEndpoint: String,
    val httpEndpoint: String,
)

data class SupportTopology(val channelEndpoint: String, val routerEndpoint: String)

data class SampleSessionNode(val routerEndpoint: String, val streamEndpoint: String)
